### ISSUE TITLE
BarrelUtils.SetBaseCachePath("Path") does not work as expected.

### ISSUE THREAD

**Issue by JMan7777** [jamesmontemagno/monkey-cache#94]
Hi,

I have a very strange behavior when I try to use the [persisting-data-longer](https://github.com/jamesmontemagno/monkey-cache#persisting-data-longer) functionality.

1. I get my "Path" value I want to use: `Application.Context?.GetExternalFilesDir("")?.AbsolutePath`
This will resolve in my example to `/storage/emulated/0/Android/data/myapppackage/files`.

2. Before I do anything with the BarrelCache I call: `BarrelUtils.SetBaseCachePath("/storage/emulated/0/Android/data/myapppackage/files");`

3. I set the `Barrel.ApplicationId` to a value. E.g. `test`.

4. This means that the cache index (idx.dat) and all files should be generated at:
`/storage/emulated/0/Android/data/myapppackage/files/test/MonkeyCacheFS`

**Now the crazy part:** 

When my app starts for the very first time (fresh install) i see that monkey cache writes files to the newly automatically by MonkeyCache generated directory `/storage/emulated/0/Android/data/myapppackage/files/test/MonkeyCacheFS`. E.g.:
```
-rw-rw---- 1 u0_a425 sdcard_rw      7 2020-11-25 11:51 04-A1-15-BD-D8-0B-87-26-80-1A-2F-C6-77-D3-FF-46
-rw-rw---- 1 u0_a425 sdcard_rw      4 2020-11-25 11:51 F0-E3-CB-57-0D-63-F9-CE-BF-AD-84-03-E5-CB-D0-D1
-rw-rw---- 1 u0_a425 sdcard_rw      4 2020-11-25 11:52 BE-0A-CA-EC-27-5D-49-50-B8-2E-6C-A5-8C-49-3F-25
-rw-rw---- 1 u0_a425 sdcard_rw   2316 2020-11-25 12:00 9E-10-49-47-42-CC-D0-09-46-D1-57-B9-89-7B-22-80
-rw-rw---- 1 u0_a425 sdcard_rw   1724 2020-11-25 12:00 B9-CC-F5-9E-4D-58-83-3E-57-A3-8B-99-33-26-CF-C8
-rw-rw---- 1 u0_a425 sdcard_rw 359650 2020-11-25 12:00 76-BB-41-D2-03-19-50-D8-83-A9-23-96-95-46-1F-8C
-rw-rw---- 1 u0_a425 sdcard_rw    247 2020-11-25 12:00 idx.dat
-rw-rw---- 1 u0_a425 sdcard_rw 427087 2020-11-25 12:00 E6-DA-4F-30-36-B7-8E-C0-7F-6C-1D-5A-A8-AE-02-E0
```
I also see that the `idx.dat` file contains the cache keys I added correctly.

Now when I fully close my app (no process anymore) and start it up again and then add some more data to the cache it 
writes the cache somewhere else and I don't know where.
The cache from an application perspective is consistent (meaning even after fully restarting the app)  but the files under `/storage/emulated/0/Android/data/myapppackage/files/test/MonkeyCacheFS` never change. Also any newly added cache keys are not in the `/storage/emulated/0/Android/data/myapppackage/files/test/MonkeyCacheFS/idx.dat` file. 
But they must be somewhere stored as the cache is still consistent after restarting the app.

I found this issue since I needed to migrate the cache folder as described [here](https://github.com/jamesmontemagno/monkey-cache/issues/93).
I basically moved all files from `/data/user/0/myapppackage/cache/test/MonkeyCacheFS` to `/storage/emulated/0/Android/data/myapppackage/files/test/MonkeyCacheFS` using normal Xamarin file copy functions and then deleted the old directory `/data/user/0/myapppackage/cache/test/MonkeyCacheFS` but when my app started it behaved as the cache was empty
(even `BarrelUtils.SetBaseCachePath("/storage/emulated/0/Android/data/myapppackage/files");` was called and  `Barrel.ApplicationId` was set.)

I then did a bit testing and found the issue as described above.
To me it looks as when MonkeyCache generates the cache directory and file structure the first time, it is working fine.
However when the cache directory and file structure is already there, it's not using/updating it.

Below is the piece of code in my app to initialize the MonkeyCache and do a migration of the cache files if needed.
However, at the moment the files get correctly migrated but MonkeyCache is not using the migrated files and stores it's data somewhere I don't know.
```
            string MyApplicationId = "test";
            try
            {
                Logger.Debug("Monkey Cache - Check if migration is needed.");
                string cacheDirectory = Application.Context?.CacheDir?.AbsolutePath;
                string storageDirectory = Application.Context?.GetExternalFilesDir("")?.AbsolutePath;

                if (cacheDirectory != null && storageDirectory != null)
                {
                    string cacheFileDirectory = Path.Combine(cacheDirectory, MyApplicationId, "MonkeyCacheFS");
                    string storageFileDirectory = Path.Combine(storageDirectory, MyApplicationId, "MonkeyCacheFS");
                    if (Directory.Exists(storageFileDirectory))
                    {
                        Logger.Debug($"Monkey Cache storage directory '{storageFileDirectory}' already exists. No migration needed.");
                    }
                    else
                    {
                        if (Directory.Exists(cacheFileDirectory))
                        {
                            Logger.Debug($"Monkey Cache cache directory '{cacheFileDirectory}' found. Will migrate to storage directory '{storageFileDirectory}'.");
                            Logger.Debug($"Monkey Cache - Create new storage directory '{storageFileDirectory}'.");
                            Directory.CreateDirectory(storageFileDirectory);

                            try
                            {
                                string[] cacheFileNames = Directory.GetFiles(cacheFileDirectory);
                                if (cacheFileNames != null && cacheFileNames.Any())
                                {
                                    cacheFileNames.ToList().ForEach(cacheFile =>
                                    {
                                        string storageFile = Path.Combine(storageFileDirectory, Path.GetFileName(cacheFile));
                                        Logger.Debug($"Monkey Cache - Copy file '{cacheFile}' to '{storageFile}'.");
                                        File.Copy(cacheFile, storageFile);
                                    });
                                }
                                else
                                {
                                    Logger.Warning($"Monkey Cache - No files found in cache directory '{cacheFileDirectory}' to be migrated.");
                                }
                            }
                            catch (Exception ex)
                            {
                                Logger.Error($"Monkey Cache - Error occurred while copying files from '{cacheFileDirectory}' to '{storageFileDirectory}'.");
                                Logger.Debug($"Monkey Cache - Deleting newly created storage directory '{storageFileDirectory}' since the migration failed.");
                                Directory.Delete(storageFileDirectory, true);
                                throw ex;
                            }
                            Logger.Debug($"Monkey Cache - Delete old cache directory '{cacheFileDirectory}'.");
                            Directory.Delete(cacheFileDirectory, true);
                            Logger.Debug($"Monkey Cache file migration done.");
                        }
                        else
                        {
                            Logger.Debug($"Monkey Cache cache directory '{cacheFileDirectory}' and storage directory '{storageFileDirectory}' are not found. No migration needed.");
                        }
                    }
                }
                else
                {
                    Logger.Warning("Could not get cache and/or storage directory to check if cache file migration is needed.");
                }

                if (storageDirectory != null)
                {
                    BarrelUtils.SetBaseCachePath(storageDirectory);
                    Logger.Debug($"Monkey Cache base storage directory defined as '{storageDirectory}'.");
                }
            }
            catch(Exception ex)
            {
                Logger.Error($"InitializeCustomBarrelCache failed with error: {ex.Message}");
            }

            Barrel.ApplicationId = MyApplicationId ;
```

**Comment by jamesmontemagno** [comment]
Hmm tricky one I think maybe you need to set the application id first... but idk shouldn't matter..... Appreciate the input here. Instead of using the statis barrel, can you call create on it after setting the application id?->
https://github.com/jamesmontemagno/app-ac-islandtracker/blob/master/TurnipTracker/Services/DataService.cs#L48-L49

**Comment by JMan7777** [comment]
Hi,

I made some progress and did actually a mistake in my code.

First I changed the way i get the directories:
```
string cacheDirectory = FileSystem.CacheDirectory;
string storageDirectory = FileSystem.AppDataDirectory;
```

This already solved the issue when I was testing in the emulator (with an Android 11 image).
So in the emulator (emulator version 30.2.6) the migration works fine and as expected.

I then built a version I ran on a physical test phone (Xiaomi - Android 10 - not rooted).
I see that that all the MonkeyCache files are correctly moved and when calling
`Barrel.Current.Exists(<existing cache key>)` after the migration
it returns `true`.

**Now the new issue:**

Even the cache key exists, when I try to get the value for the key via:
`Barrel.Current.Get<T>(key)`
it's always null.

Again all the files cache files were copied, but for some reason the value for a key can't be retrieved and is null.
When I add new entries to the cache after the migration, those values are fine.
Only migrated values are always null.

Any idea ?
